File: javascript/ql/test/query-tests/Security/CWE-312/passwords.js
(function() {
    console.log(password); // $ Alert[js/clear-text-logging]
    console.log(o.password); // $ Alert[js/clear-text-logging]
    console.log(getPassword()); // $ Alert[js/clear-text-logging]
    console.log(o.getPassword()); // $ Alert[js/clear-text-logging]

    function myLog(x) {
        console.log(x); // $ Alert[js/clear-text-logging]
    }
    myLog(password); // $ Source[js/clear-text-logging]

    console.info(password); // $ Alert[js/clear-text-logging]

    console.log(name + ", " + password); // $ Alert[js/clear-text-logging]

    console.log(`${name}, ${password}`); // $ Alert[js/clear-text-logging]

    var obj1 = {
        password: x // $ Source[js/clear-text-logging]
    };
    console.log(obj1); // $ Alert[js/clear-text-logging]

    var obj2 = {
        x: password // $ Source[js/clear-text-logging]
    };
    console.log(obj2); // $ Alert[js/clear-text-logging]

    var obj3 = {};
    console.log(obj3);
    obj3.x = password;

    var fixed_password = "123";
    console.log(fixed_password);

    console.log(messages.IncorrectPasswordError);

    console.log(this.hashed_password);
    console.log(login.wrappedJSObject.encryptedPassword);
    console.log(HTML5QQ.encodedPassword);

    console.log({password: crypt(pw)});
    var actually_secure_password = crypt(password);
    console.log(actually_secure_password);

    var user1 = {};
    user1.crypted_password = x();
    console.log(user1);

    var user2 = {};
    user2.password = hash();
    console.log(user2);

    var user3 = {
        password: encryptLib.encryptPassword(req.body.password)
    };
    console.log(user3);

    var actually_ok_password_1 = hashed1();
    console.log(actually_ok_password_1);
    var hashed2 = actually_ok_password_2();
    console.log(hashed2);

    console.log(this._passwordMD5);
    console.log(this._password_sha);

    var utilityObject = {
        passwordSet: function(){}
    }
    console.log(utilityObject);

    var displayedPassword = password? null: "(hidden)";
    var config = {
        password: displayedPassword
    };
    console.log(config);

    var temp = { encryptedPassword: req.body.password }; // $ Source[js/clear-text-logging]
    console.log(temp.encryptedPassword); // $ SPURIOUS: Alert[js/clear-text-logging]

    var secret = password; // $ Source[js/clear-text-logging]
    console.log(`pw: ${secret}`); // $ Alert[js/clear-text-logging]

    function redact(kind, value) {
        if (value && kind == 'password') {
            return "********"
        }
        return value;
    }

    console.log("Password is: " + redact('password', password));

    if (environment.isTestEnv()) {
        console.log("Password is: " + password); // $ SPURIOUS: Alert[js/clear-text-logging]
    }

    if (environment.is(TEST)) {
        // NB: for security reasons, we only log passwords in test environments
        console.log("Password is: " + password); // $ SPURIOUS: Alert[js/clear-text-logging]
    }


    if (x.test(y)) {
        f();
        // ...
        console.log("Password is: " + password); // $ Alert[js/clear-text-logging]
        // ...
    }

    if (environment.isTestEnv())
        console.log("Password is: " + password); // $ SPURIOUS: Alert[js/clear-text-logging]

    if (x.test(y)) {
        if (f()) {
            console.log("Password is: " + password); // $ Alert[js/clear-text-logging]
        }
    }

    if (!environment.isProduction()) {
        console.log("Password is: " + password); // $ SPURIOUS: Alert[js/clear-text-logging]
    }

    console.log(name + ", " + password.toString()); // $ Alert[js/clear-text-logging]
    console.log(name + ", " + password.valueOf()); // $ Alert[js/clear-text-logging]
});

(function() {
    var config = {
        password: x, // $ Source[js/clear-text-logging]
        hostname: "tarski",
        x: password, // $ Source[js/clear-text-logging]
        y: getPassword() // $ Source[js/clear-text-logging]
    };
    var cfg = x? config: config;
    console.log(config.hostname);
    console.log(config); // $ Alert[js/clear-text-logging]
    console.log(config.x); // $ Alert[js/clear-text-logging]
    console.log(config.y); // $ Alert[js/clear-text-logging]
    console.log(config[x]); // OK - probably
});

function indirectLogCall() {
	console.log.apply(this, arguments); // $ Alert[js/clear-text-logging]
}
var Util = require('util');
(function() {
    var config = {
        x: password // $ Source[js/clear-text-logging]
    };
    indirectLogCall(config.x);
    indirectLogCall(process.env); // $ Source[js/clear-text-logging]

    var procdesc = Util.inspect(process.env).replace(/\n/g, '') // $ Source[js/clear-text-logging]

    indirectLogCall(procdesc);

    console.log(process.env); // $ Alert[js/clear-text-logging]
    console.log(process.env.PATH);
    console.log(process.env["foo" + "bar"]);
});

(function () {
	console.log(password.replace(/./g, "*"));
	console.log(password.replace(/\./g, "*")); // $ Alert[js/clear-text-logging]
	console.log(password.replace(/foo/g, "*")); // $ Alert[js/clear-text-logging]
})();

const debug = require('debug')('test');
(function () {
    console.log(password.replace(/foo/g, "*")); // $ Alert[js/clear-text-logging]
    debug(password.replace(/foo/g, "*")); // $ Alert[js/clear-text-logging]

    const myPassword = foo();
    console.log(myPassword); // $ Alert[js/clear-text-logging]

    const myPasscode = foo();
    console.log(myPasscode); // $ Alert[js/clear-text-logging]
});

(function () {
    console.log(password.replace(/./g, "*"));
	console.log(password.replace(new RegExp(".", "g"), "*"));
	console.log(password.replace(new RegExp("."), "*")); // $ Alert[js/clear-text-logging]
	console.log(password.replace(new RegExp(".", unknownFlags()), "*")); // OK - Most likely not a problem.
    console.log(password.replace(new RegExp("pre_._suf", "g"), "*"));
})();

(function () {
  console.log(JSON.stringify(process.env)); // $ Alert[js/clear-text-logging]
  console.log(process.env.PATH);
});

File: go/ql/test/query-tests/Security/CWE-312/passwords.go
package main

import (
	"log"
	"os"
)

func myLog(x string) {
	log.Println(x) // $ Alert
}

func redact(kind, value string) string {
	if value != "" && kind == "password" {
		return "********"
	}
	return value
}

func test() {
	name := "user"
	password := "P@ssw0rd" // $ Source
	x := "horsebatterystapleincorrect"
	var o passStruct

	log.Println(password)        // $ Alert
	log.Println(o.password)      // $ Alert
	log.Println(getPassword())   // $ Alert
	log.Println(o.getPassword()) // $ Alert

	myLog(password)

	log.Panic(password) // $ Alert

	log.Println(name + ", " + password) // $ Alert

	obj1 := passStruct{
		password: x, // $ Source
	}
	log.Println(obj1) // $ Alert

	obj2 := xStruct{
		x: password,
	}
	log.Println(obj2) // $ Alert

	var obj3 xStruct
	log.Println(obj3)
	obj3.x = password

	fixed_password := "cowbatterystaplecorrect" // $ Source
	log.Println(fixed_password)                 // $ Alert // Probably OK

	log.Println(IncorrectPasswordError) // OK

	var obj hashedStruct
	log.Println(obj.hashed_password) // OK
	var login encryptedStruct
	log.Println(login.encryptedPassword) // OK
	var HTML5QQ encodedStruct
	log.Println(HTML5QQ.encodedPassword) // OK

	pw := "arstneioarstneio"
	log.Println(passStruct{password: crypt(pw)}) // OK
	actually_secure_password := crypt(password)  // OK
	log.Println(actually_secure_password)        // OK

	var user1 cryptedStruct
	x2 := "perhaps sensitive"
	user1.cryptedPassword = x2
	log.Println(user1) // OK

	var user2 passStruct
	user2.password = hash()
	log.Println(user2) // OK

	actually_ok_password_1 := hash()
	log.Println(actually_ok_password_1) // OK
	hashed2 := actually_ok_password_2()
	log.Println(hashed2) // OK

	passwordMD5 := "0d599f0ec05c3bda8c3b8a68c32a1b47"
	log.Println(passwordMD5) // OK
	password_sha := "5e03aad954f62810c0fdf16cb0dfba86bda486a53912fb7c894f96eaf8008aee"
	log.Println(password_sha) // OK

	utilityObject := passSetStruct{
		passwordSet: make(map[string]bool), // $ Source
	}
	log.Println(utilityObject) // $ Alert

	secret := password
	log.Printf("pw: %s", secret) // $ Alert

	log.Println("Password is: " + redact("password", password))

	var t testable
	var y int

	if t.test(y) {
		f()
		// ...
		log.Println("Password is: " + password) // $ Alert
		// ...
	}

	if t.test(y) {
		if f() {
			log.Println("Password is: " + password) // $ Alert
		}
	}

	if os.Getenv("APP_ENV") != "production" {
		log.Println("Password is: " + password) // $ SPURIOUS: Alert
	}

	var password1 stringable = stringable{"arstneio"} // $ Source
	log.Println(name + ", " + password1.String())     // $ Alert

	x3 := "sheepbatterystaplecorrect"
	config := Config{
		password: x3, // $ Source
		hostname: "tarski",
		x:        password,
		y:        getPassword(), // $ Source
	}
	log.Println(config.hostname) // OK
	log.Println(config)          // $ Alert
	log.Println(config.x)        // $ Alert
	log.Println(config.y)        // $ Alert

	obj4 := xStruct{
		x: "aaaaa",
	}
	log.Println(obj4) // OK
}

const password = "horsebatterystaplecorrect"

func test2() {
	log.Println(password) // OK
}
